ml-agents/mlagents/trainers/torch/components/reward_providers/gail_reward_provider.py

```diff
@@ -220,10 +218,28 @@ def compute_gradient_magnitude(
         Gradient penalty from https://arxiv.org/pdf/1704.00028. Adds stability esp.
         for off-policy. Compute gradients w.r.t randomly interpolated input.
         """
-        policy_obs = self.get_state_encoding(policy_batch)
-        expert_obs = self.get_state_encoding(expert_batch)
-        obs_epsilon = torch.rand(policy_obs.shape)
-        encoder_input = obs_epsilon * policy_obs + (1 - obs_epsilon) * expert_obs
+        policy_vec_inputs, policy_vis_inputs = self.get_state_inputs(policy_batch)
+        expert_vec_inputs, expert_vis_inputs = self.get_state_inputs(expert_batch)
+        interp_vec_inputs = []
+        for policy_vec_input, expert_vec_input in zip(
+            policy_vec_inputs, expert_vec_inputs
+        ):
+            obs_epsilon = torch.rand(policy_vec_input.shape)
+            interp_vec_input = (
+                obs_epsilon * policy_vec_input + (1 - obs_epsilon) * expert_vec_input
+            )
+            interp_vec_input.requires_grad = True  # For gradient calculation
+            interp_vec_inputs.append(interp_vec_input)
+        interp_vis_inputs = []
+        for policy_vis_input, expert_vis_input in zip(
+            policy_vis_inputs, expert_vis_inputs
+        ):
+            obs_epsilon = torch.rand(policy_vis_input.shape)
+            interp_vis_input = (
+                obs_epsilon * policy_vis_input + (1 - obs_epsilon) * expert_vis_input
+            )
+            interp_vis_input.requires_grad = True  # For gradient calculation
```

Comment:
I do not think we need to require a gradient going into the inputs.

Reply:
This is required. As I understand it, torch wont track gradients with respect to any variables where this is not set to true.  https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html#tensor

This code errors with this 
`RuntimeError: One of the differentiated Tensors does not require grad`
if this is not true.